build-tools/packages/build-tools/src/typeValidator/packageJson.ts

```diff
@@ -72,69 +80,172 @@ export async function getPackageDetails(packageDir: string): Promise<PackageDeta
     };
 }
 
+type PreviousVersionStyle =
+    | "baseMinor"
+    | "baseMajor"
+    | "^previousMajor"
+    | "^previousMinor"
+    | "~previousMajor"
+    | "~previousMinor"
+    | "previousMajor"
+    | "previousMinor";
+
+/**
+ * Based on the current version of the package as per package.json, determines the previous version that we should run
+ * typetests against.
+ *
+ * The version used for the previous version can be adjusted by passing different "style" values in via the
+ * previousVersionStyle parameter.
+ *
+ * @param packageDir - The path to the package.
+ * @param writeUpdates - If true, will update the package.json with new previous versions.
+ * @param previousVersionStyle - The version style to use when determining the previous version. Can be the exact
+ * previous major or minor versions, or caret/tilde-equivalent dependency ranges on those previous versions.
+ * @param exactPreviousVersionString - If provided, this string will be used as the previous version string.
+ * @param resetBroken - If true, clears the "broken" section of the type validation, effectively clearing all known
+ * breaking changes.
+ * @returns package metadata or a reason the package was skipped.
+ */
 export async function getAndUpdatePackageDetails(
     packageDir: string,
-    updateOptions: { cwd?: string } | undefined,
+    writeUpdates: boolean | undefined,
+    previousVersionStyle: PreviousVersionStyle = "baseMinor",
+    exactPreviousVersionString?: string,
+    resetBroken?: boolean,
 ): Promise<(PackageDetails & { skipReason?: undefined }) | { skipReason: string }> {
     const packageDetails = await getPackageDetails(packageDir);
 
     if (packageDetails.pkg.name.startsWith("@fluid-internal")) {
-        return { skipReason: "Skipping package: @fluid-internal " };
+        return { skipReason: "Skipping package: @fluid-internal" };
     } else if (packageDetails.pkg.main?.endsWith("index.js") !== true) {
         return { skipReason: "Skipping package: no index.js in main property" };
     } else if (packageDetails.pkg.private === true) {
         return { skipReason: "Skipping package: private package" };
+    } else if (packageDetails.pkg.typeValidation === undefined) {
+        return {
+            skipReason: 'Skipping package: add "typeValidation: {}" to package.json to enable.',
+        };
     } else if (packageDetails.pkg.typeValidation?.disabled === true) {
         return { skipReason: "Skipping package: type validation disabled" };
```

Comment:
If we add new package (say client package), can we enforce type validation?

Reply:
Before this PR, typetest preparation was run on a folder basis, and ran only on the "packages" folder. The previous code always added typetests to all the packages found in/under that folder.

With this PR, generation is run on a whole release group. Making type test generation "opt in" was faster/easier than disabling it in all the examples and experimental stuff. That said, I can be convinced that is a better, more accurate change. What do you think?

Reply:
I would preferable stay closer to the safer side. It could be enrollment per release group (seems very reasonable, but Im not sure how it plays out with examples) or folder, but once enrolled, enforcing typetests would be a way to ensure some package does not sneak in and simply ignores the rule. Or if it wants to bypass typetests, it explicitly states that. Option could be some explict opt out per folder, and never allow individual package to opt out individually.

This is all from perspective "I trust what typetests do", and you adding pinning alternative in short term very much helps that :)

Reply:
I'll revert this change and opt out individual packages as needed when making the actual changes.